### PR TITLE
Feature/move shard fix

### PR DESCRIPTION
**Pull request by fceller**

### CODE REVIEW

File: arangod/Cluster/ClusterComm.cpp [hunk 1]
```diff
@@ -942,7 +943,15 @@ size_t ClusterComm::performRequests(std::vector<ClusterCommRequest>& requests,
           << requests[index].path << " with return code "
           << (int)res.answer_code;
       } else if (res.status == CL_COMM_BACKEND_UNAVAILABLE ||
-         (res.status == CL_COMM_TIMEOUT && !res.sendWasComplete)) {
+                 (res.status == CL_COMM_TIMEOUT && !res.sendWasComplete) ||
+                 (res.status == CL_COMM_RECEIVED &&
+                  res.answer_code == rest::ResponseCode::NOT_ACCEPTABLE)) {
+        // Note that this case includes the refusal of a leader to accept
+        // the operation, in which we have to flush ClusterInfo:
+        if (res.status == CL_COMM_RECEIVED &&
```
Comment (fceller) [inline arangod/Cluster/ClusterComm.cpp]: Do not retry for DB->Db, but fail instead

File: arangod/Transaction/Methods.cpp [hunk 2]
```diff
@@ -1469,7 +1497,7 @@ OperationResult transaction::Methods::insertLocal(
         "/_db/" + arangodb::basics::StringUtils::urlEncode(databaseName()) +
         "/_api/document/" +
         arangodb::basics::StringUtils::urlEncode(collection->name()) +
-        "?isRestore=true";
+        "?isRestore=true&isSynchronousReplication=true";
```
Comment (fceller) [inline arangod/Transaction/Methods.cpp]: Needs leader name

File: arangod/Transaction/Methods.cpp [hunk 3]
```diff
@@ -1518,7 +1546,17 @@ OperationResult transaction::Methods::insertLocal(
...
         if (nrGood < followers->size()) {
-          // we drop all followers that were not successful:
+          // If any would-be-follower refused to follow there must be a
+          // new leader in the meantime, in this case we must not allow
+          // this operation to succeed, we simply return with a refusal
+          // error (note that we use the follower version, since we have
+          // lost leadership):
+          if (findRefusal(requests)) {
+            return OperationResult(
+                TRI_ERROR_CLUSTER_SHARD_FOLLOWER_REFUSES_OPERATION);
```
Comment (fceller) [inline arangod/Transaction/Methods.cpp]: 3rd error code: LEADER RESIGNED

File: arangod/V8Server/v8-collection.cpp [hunk 4]
```diff
@@ -1044,10 +1059,10 @@ static void JS_FiguresVocbaseCol(
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-/// @brief was docuBlock leaderResign
+/// @brief was docuBlock assumeLeadership
 ////////////////////////////////////////////////////////////////////////////////
 
-static void JS_LeaderResign(v8::FunctionCallbackInfo<v8::Value> const& args) {
+static void JS_SetLeader(v8::FunctionCallbackInfo<v8::Value> const& args) {
```
Comment (fceller) [inline arangod/V8Server/v8-collection.cpp]: SetTheLeader

File: arangod/V8Server/v8-collection.cpp [hunk 5]
```diff
@@ -1076,63 +1091,32 @@ static void JS_LeaderResign(v8::FunctionCallbackInfo<v8::Value> const& args) {
...
+    if (args.Length() >= 1 && args[0]->IsString()) {
+      TRI_Utf8ValueNFC l(TRI_UNKNOWN_MEM_ZONE, args[0]);
+      theLeader = std::string(*l, l.length());
     }
-
-    std::string collectionName = v8Collection->name();
-    auto collection = vocbase->lookupCollection(collectionName);
-    if (collection == nullptr) {
-      TRI_V8_THROW_EXCEPTION(TRI_ERROR_ARANGO_COLLECTION_NOT_FOUND);
+    collection->followers()->setLeader(std::string(""));
```
Comment (fceller) [inline arangod/V8Server/v8-collection.cpp]: needs theLeader

File: js/server/modules/@arangodb/cluster.js [hunk 6]
```diff
@@ -359,7 +359,7 @@ function organiseLeaderResign (database, collId, shardName) {
   try {
     // we know the shard exists locally!
     var db = require('internal').db;
-    db._collection(shardName).leaderResign();
+    db._collection(shardName).setLeader("");  // resign
```
Comment (fceller) [inline js/server/modules/@arangodb/cluster.js]: "" = I'm leader

File: arangod/Cluster/ClusterComm.cpp (hunk 1 above)
Comment (neunhoef) [inline arangod/Cluster/ClusterComm.cpp]: Done

File: arangod/Transaction/Methods.cpp (hunk 2 above)
Comment (neunhoef) [inline arangod/Transaction/Methods.cpp]: Done

File: arangod/Transaction/Methods.cpp (hunk 3 above)
Comment (neunhoef) [inline arangod/Transaction/Methods.cpp]: Done

File: arangod/V8Server/v8-collection.cpp (hunk 4 above)
Comment (neunhoef) [inline arangod/V8Server/v8-collection.cpp]: Done

File: arangod/V8Server/v8-collection.cpp (hunk 5 above)
Comment (neunhoef) [inline arangod/V8Server/v8-collection.cpp]: Done

File: js/server/modules/@arangodb/cluster.js (hunk 6 above)
Comment (neunhoef) [inline js/server/modules/@arangodb/cluster.js]: Done

**Pull request merged**